File: linked_list_cycle/linked_list_cycle.py
# MY BRUTE FORCE SOLUTION
def hasCycle(self, head: Optional[ListNode]) -> bool:
    history = []
    index = 0
    if not head:
        return False

    def seen(node, history):
        if node in history:
            return history.index(node)
        else:
            if node.next == None:
                return -1
            else:
                new_history = history + [node]
                return self.seen(node.next, new_history)

    return seen(head, history) >= 0

# OPTIMAL SOLUTION
def hasCycleOptimal(self, head: Optional[ListNode]) -> bool:
    if not head:
        return False
    slow = fast = head
    while fast and fast.next:
        fast = fast.next.next
        slow = slow.next
        if slow == fast:
            return True
    return False

File: leet_code_algorithms/linked_list_cycle/linked_list_cycle.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

const hasCycle = (head: ListNode | null): boolean => {
  if (!head) {
    return false;
  }
  let slow: ListNode | null = head;
  let fast: ListNode | null = head;
  while (fast && fast.next) {
    slow = slow!.next;
    fast = fast.next.next;
    if (slow === fast) {
      return true;
    }
  }
  return false;
};

var hasCycleUsingMap = function (head) {
  //previous one was way too brute, using less brute now
  if (!head) {
    return false;
  }
  let curr = head;
  let nodeMap = new Map();
  while (curr) {
    if (nodeMap.has(curr)) {
      return true;
    }
    nodeMap.set(curr, true);
    curr = curr.next;
  }
  return false;
};
